Add the installer

Vincere1st opened the pull request:
I'm created the installer and tested.

I guess the next step after review and merge on development, is to find users to test this installer.

And if all is ok, merge to main.

Commit:
feat(install): add Docker and Docker Compose and DDB with all configuration

Commit:
feat(install): check if the script is run with sudo

Commit:
feat(docker): improvement script

Commit:
feat(install): replace ~ by $HOME

Commit:
feat(install): automatic acceptation for apt

Commit:
fix(script): change how write into files override.conf and daemon.json

Commit:
feat(install): add make

Commit:
fix(installer-ddb): add DDB_USER for run ddb installer as non-root thks Jérémy

Commit:
feat(install-ddb): add configuration to ddb.yml and add global alias

Commit:
add execute flag and change path directory

Commit:
feat(installer): add the update of alias

Commit:
feat(docs): add the README file

jeremyVignelles commented on README.md:
```diff
@@ -0,0 +1,24 @@
+WSL2 with Ubuntu
+
+## Installation
+Clone this project and run the following command in the project directory:
```
```suggestion
Clone this project inside your WSL and run the following command in the project directory:
```

jeremyVignelles replied:
Also put the git clone command here, people are lazy.

Vincere1st replied:
Corrected

jeremyVignelles commented on README.md:
```diff
@@ -0,0 +1,24 @@
+And you don't even have to install Docker Desktop and therefore without license limitation linked to Docker Desktop.
+
+## Requirements
+WSL2 with Ubuntu
```
with git installed

Vincere1st replied:
Normally git is installed with Ubuntu, I didn't install it to clone this project.

Vincere1st replied:
corrected

jeremyVignelles commented on install.sh:
```diff
@@ -0,0 +1,100 @@
+DDB_USER=${DDB_USER:-$SUDO_USER}
+
+if [ -z "$DDB_USER" ]; then
+    echo "DDB_USER is not defined"
```
```suggestion
    echo "DDB_USER is not defined or could not be detected from SUDO_USER. Set the DDB_USER variable, or run this script as your target user (must be in sudoer)."
```

Vincere1st replied:
ocrrected

jeremyVignelles commented on install.sh:
```diff
@@ -0,0 +1,100 @@
+
+# Check if the script is run as root if not rerun as root
+if [ "$(id -u)" != "0" ]; then
+   sudo $0
+    exit $?
```
Those lines are not aligned.

Vincere1st replied:
corrected

jeremyVignelles commented on install.sh:
```diff
@@ -0,0 +1,100 @@
+
+#Install and configuration of Docker
+## Add Docker's official GPG key:
+apt-get update -y
```
You've litterally just done it, no need to do it again.

Vincere1st replied:
you're right, corrected

jeremyVignelles commented on install.sh:
```diff
@@ -0,0 +1,100 @@
+usermod -aG docker $DDB_USER
+
+## Change permissions of the Docker socket
+chmod 666 /var/run/docker.sock
```
Not sure we need this one. What is it for ?

Vincere1st replied:
if we don't do this we have permissions issue. Example with a docker compose ps:
`permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.44/containers/json?filters=%7B%22label%22%3A%7B%22com.docker.compose.config-hash%22%3Atrue%2C%22com.docker.compose.oneoff%3DFalse%22%3Atrue%2C%22com.docker.compose.project%3Dtraefik%22%3Atrue%7D%7D": dial unix /var/run/docker.sock: connect: permission denied`

after run the chmod all is ok

jeremyVignelles replied:
should be enough to just refresh the user's group with `newgrp docker`
source: https://docs.docker.com/engine/install/linux-postinstall/

jeremyVignelles commented on install.sh:
```diff
@@ -0,0 +1,100 @@
+systemctl restart docker
+
+# Install and configuration of Docker Compose
+DOCKER_CONFIG=${DOCKER_CONFIG:-/home/$DDB_USER/.docker}
+mkdir -p $DOCKER_CONFIG/cli-plugins
+curl -SL https://github.com/docker/compose/releases/download/v2.20.3/docker-compose-linux-x86_64 -o $DOCKER_CONFIG/cli-plugins/docker-compose
+# Apply executable permissions to the binary
+chmod +x $DOCKER_CONFIG/cli-plugins/docker-compose
```
Isn't that what the docker-compose-plugin on line 36 was for ?
Also not a fan of the hardcoded version number.

jeremyVignelles commented on install.sh:
```diff
@@ -0,0 +1,100 @@
+    host: '\$HOST_IP'
+EOF
+
+# Configure the global ddb.local.yml file
+cat <<EOF > "/home/$DDB_USER/.docker-devbox/ddb.local.yaml"
+certs:
+  cfssl:
+    server:
+      host: inetum-cfssl.azurewebsites.net
+      port: 443
+      ssl: true
+      verify_cert: true
+shell:
+  aliases:
+    dc: docker compose
+  global_aliases:
+    - dc
+EOF
```
This should be split in two.
The dc alias can be moved into ddb.yaml, but the certs part is really inetum-specific. It's most likely a post-setup task that we can't automate with this script (as well as the azure credentials...)

jeremyVignelles replied:
We can merge the shell config into ddb.yaml, and leave the certs part for #4

jeremyVignelles commented on install.sh:
```diff
@@ -0,0 +1,100 @@
+EOF
+
+# Reload the .bashrc file
+source /home/$DDB_USER/.bashrc
```
Probably doesn't work, as we are in a sudo session...

Commit:
feat(docs): added 2 first suggested comments in the pull resquest review

Commit:
feat(installer): add more description error and remove the double apt update

Vincere1st commented on install.sh:
```diff
@@ -0,0 +1,99 @@
+echo "[Service]\nExecStart=\nExecStart=/usr/bin/dockerd -H tcp://0.0.0.0:2375 -H unix:///var/run/docker.sock" > /etc/systemd/system/docker.service.d/override.conf
+
+## Create daemon.json file to set default address pool
+echo "{\n \"default-address-pools\": [{\"base\":\"10.199.0.0/16\", \"size\":24}]}" > /etc/docker/daemon.json
```
The "\n" is write in the file: fix this

Vincere1st commented on install.sh:
```diff
@@ -0,0 +1,99 @@
+mkdir -p /etc/systemd/system/docker.service.d
+
+## Create override.conf file to expose docker to Windows
+echo "[Service]\nExecStart=\nExecStart=/usr/bin/dockerd -H tcp://0.0.0.0:2375 -H unix:///var/run/docker.sock" > /etc/systemd/system/docker.service.d/override.conf
```
The "\n" is write in the file: fix this ( 2 x "\n" )

jeremyVignelles commented on README.md:
````diff
@@ -0,0 +1,28 @@
+WSL2 with Ubuntu and check if git is installed, if not install it.
+
+## Installation
+Clone this project inside your WSL and run the following command in the project directory:
+```bash
+git clone https://github.com/inetum-orleans/docker-devbox-wsl.git
+
+cd docker-devbox-wsl
+
+./install.sh
+```
````
Note for later (because we need to ensure this works first), we could rewrite it like this instead:

```suggestion
Run this command to download the script and run it as root:
```bash
curl https://raw.githubusercontent.com/inetum-orleans/docker-devbox-wsl/main/install.sh | sudo bash
```
```

jeremyVignelles commented on install.sh:
```diff
@@ -0,0 +1,99 @@
+apt-get update -y
+apt-get upgrade -y
+
+#Install and configuration of Docker
+## Add Docker's official GPG key:
+apt-get install ca-certificates curl make -y
+install -m 0755 -d /etc/apt/keyrings
+curl -fsSL https://download.docker.com/linux/ubuntu/gpg -o /etc/apt/keyrings/docker.asc
+chmod a+r /etc/apt/keyrings/docker.asc
+
+## Add the repository to Apt sources:
+echo \
+  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu \
+  $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \
+  tee /etc/apt/sources.list.d/docker.list > /dev/null
+
+apt-get update -y
+
+apt-get install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin -y
```
Can probably be replaced with the script from get.docker.com to support debian

Vincere1st commented:
~This issue #5 must be corrected before merging~ **Done**

Commit:
fix(install): manage jump line when writing configuration file in script

Commit:
feat(installer): remove installation of docker-compose because it's a script in docker now

Commit:
feat(installer): add control if the line configuration exist before writing and add control of docker service is started

Commit:
feat(installer): add check control path and variabilisation of path

Commit:
fix(installer): activate the change to group instead of change permissions

Commit:
feat(installer): remove specific configuration of inetum

Commit:
feat(installer): remove newgrp docker and add message to the end of the script

Vincere1st committed:
feat(azure-cli): azure-cli script installer

Vincere1st committed:
fix(azure-cli): Managing script termination in case of error

Vincere1st committed:
feat(azure-cli): add check if all values is not empty and auto-logging

jeremyVignelles commented on install.sh:
```diff
@@ -37,8 +37,8 @@ apt-get install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docke
-## Change permissions of the Docker socket
-chmod 666 /var/run/docker.sock
+## Activate the changes to groups
+newgrp docker
```
Not needed anymore if I remember correctly : newgrp blocks the session, but the only place where this is needed is on the ddb installer, which already uses su -c to install

jeremyVignelles commented on azure-cli.sh:
```diff
@@ -0,0 +1,37 @@
+echo "Please insert the name of azure container registry :"
+read AZ_REGISTRY
+
+echo "Please insert the username Azure :"
```
```suggestion
echo "Azure username :"
```

jeremyVignelles commented on azure-cli.sh:
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
+set -e
+
+echo "Please insert the name of azure container registry :"
```
```suggestion
echo "Azure container registry (ACR) name ( the X part in X.azurecr.io ):"
```

jeremyVignelles commented on azure-cli.sh:
```diff
@@ -0,0 +1,37 @@
+echo "Please insert the username Azure :"
+read AZ_USERNAME
+
+echo "Please insert the Tenant Azure :"
```
```suggestion
echo "Azure tenant ID :"
```

jeremyVignelles replied:
I'd move the "azure tenant" question before the username, to keep username/password next to each other.

jeremyVignelles commented on azure-cli.sh:
```diff
@@ -0,0 +1,37 @@
+echo "Please insert the Tenant Azure :"
+read AZ_TENANT
+
+echo "Please insert the password Azure :"
```
```suggestion
echo "Azure password :"
```

jeremyVignelles commented on install.sh:
```diff
@@ -0,0 +1,118 @@
+## By default, the Docker daemon requires root privileges to run. To use Docker without sudo, add your user to the Docker group
+usermod -aG docker $DDB_USER
+
+## Expose docker to Windows by 2375 port create as root
+mkdir -p /etc/systemd/system/docker.service.d
+
+## Create override.conf file to expose docker to Windows
+### Define the string to check and add
+EXPOSE_DOCKER="[Service]\nExecStart=\nExecStart=/usr/bin/dockerd -H tcp://0.0.0.0:2375 -H unix:///var/run/docker.sock"
+### Define the path to the override.conf file
+OVERRIDE_PATH="/etc/systemd/system/docker.service.d/override.conf"
+
+### Check if the file exist or the string exists in the override.conf file
+if [ ! -f "$OVERRIDE_PATH" ] || ! grep -Fxq "$EXPOSE_DOCKER" "$OVERRIDE_PATH"
+then
+    ### If the string does not exist, append it to the file
+    echo -e "$EXPOSE_DOCKER" >> "$OVERRIDE_PATH"
+fi
```
not sure we actually need that : 

tcp exposure was made as a way for PHPStorm to run docker commands (by using the docker(-compose).exe and tcp URL of the daemon. Isn't there a WSL option nowadays where it can find the docker daemon instead ?

Commit:
feat(installer): move set of environment variables

Commit:
feat(installer): remove local_ip not needed

Commit:
Update azure-cli.sh

Co-authored-by: Jérémy VIGNELLES <jeremyVignelles@users.noreply.github.com>

Commit:
feat(azure.cli): Adjust the questions form.

https://github.com/inetum-orleans/docker-devbox-wsl/pull/1#discussion_r1599506209

https://github.com/inetum-orleans/docker-devbox-wsl/pull/1#discussion_r1599507054

https://github.com/inetum-orleans/docker-devbox-wsl/pull/1#discussion_r1599508336

jeremyVignelles commented on azure-cli.sh:
```diff
@@ -0,0 +1,37 @@
+  CRONTAB_CONTENT=$(crontab -l 2>/dev/null || true)
+
+  # Test if user crontab contains `az acr login --name $AZ_REGISTRY`. If not, add it.
+  if ! echo "$CRONTAB_CONTENT" | grep -q "az acr login --name $AZ_REGISTRY"; then
```
It may return a false positive, because it looks into the crontab if it CONTAINS that string (as specified above).
But you know too well that "contains" may be tricky : In the rare case where I already have a "inetum1" registry and I want to add a "inetum" registry, it will not add it to the crontab.

Don't know if you can specify something like $ for the end of line, or maybe enclose `$AZ_REGISTRY` inside quotes ? like `az acr login --name \"$AZ_REGISTRY\"` (and adapt below).

You might also consider I'm nitpicking and ignore this message :)

jeremyVignelles committed:
chore(install): do not expose docker daemon

jeremyVignelles commented on install.sh:
```diff
@@ -0,0 +1,83 @@
+fi
+
+# Add the HOST_IP variable to the .bashrc file
+echo 'export HOST_IP=$(ip route show default | awk '\''{print $3}'\'')' >> /home/$DDB_USER/.bashrc
```
On my wsl (mirrored mode):

```sh
$ ip route show default
default dev eth6 proto kernel scope link metric 1 
default via 192.168.1.1 dev eth0 proto kernel metric 55
```

so HOST_IP is "eth6 192.168.1.1" which is wrong.

I'd really like to remove this env variable and/or use a static `172.0.0.1` "by convention" in the ddb.yml.
That 172.0.0.1 corresponds to the IP of the loopback network card I'm adding on windows, and this works for both intelliJ on windows and inside WSL with the gateway.

jeremyVignelles commented on install.sh:
```diff
@@ -0,0 +1,83 @@
+echo 'export HOST_IP=$(ip route show default | awk '\''{print $3}'\'')' >> /home/$DDB_USER/.bashrc
+
+# Configure the global ddb.yml file
+cat <<EOF > "/home/$DDB_USER/ddb.yaml"
```
this file gets created as root. Do we need a 

```sh
chown $DDB_USER:$DDB_USER /home/$DDB_USER/ddb.yaml
```
afterwards ?

jeremyVignelles commented on install.sh:
```diff
@@ -0,0 +1,83 @@
+## Add Docker's official GPG key:
+apt-get install ca-certificates curl make -y
+install -m 0755 -d /etc/apt/keyrings
+curl -fsSL https://download.docker.com/linux/ubuntu/gpg -o /etc/apt/keyrings/docker.asc
```
If docker already exist, there is no need to install it.
